src/hotspot/share/gc/shared/slidingForwarding.cpp

```diff
@@ -0,0 +1,168 @@
+    _num_regions = 1;
+    _region_size_words = round_up_power_of_2(heap_size_words);
+    _region_size_words_shift = log2i_exact(_region_size_words);
+  }
+}
+
+SlidingForwarding::~SlidingForwarding() {
+  FREE_C_HEAP_ARRAY(region_bases, _bases_table);
+  _bases_table = nullptr;
+  delete _fallback_table;
+  _fallback_table = nullptr;
+}
+
+void SlidingForwarding::begin() {
+  assert(_bases_table == nullptr, "Should be uninitialized");
+  size_t max = _num_regions * NUM_TARGET_REGIONS;
+  _bases_table = NEW_C_HEAP_ARRAY(HeapWord*, max, mtGC);
+  for (size_t i = 0; i < max; i++) {
+    _bases_table[i] = UNUSED_BASE;
+  }
+}
+
+void SlidingForwarding::end() {
+  assert(_bases_table != nullptr, "Should be initialized");
+  FREE_C_HEAP_ARRAY(HeapWord*, _bases_table);
+  _bases_table = nullptr;
+
+  delete _fallback_table;
+  _fallback_table = nullptr;
+}
+
+void SlidingForwarding::fallback_forward_to(HeapWord* from, HeapWord* to) {
+  if (_fallback_table == nullptr) {
+    _fallback_table = new FallbackTable();
+  }
+  _fallback_table->forward_to(from, to);
+}
+
+HeapWord* SlidingForwarding::fallback_forwardee(HeapWord* from) const {
+  assert(_fallback_table != nullptr, "fallback table must be present");
+  return _fallback_table->forwardee(from);
+}
+
+FallbackTable::FallbackTable() {
+  for (size_t i = 0; i < TABLE_SIZE; i++) {
+    _table[i]._next = nullptr;
+    _table[i]._from = nullptr;
+    _table[i]._to   = nullptr;
+  }
+}
+
+FallbackTable::~FallbackTable() {
+  for (size_t i = 0; i < TABLE_SIZE; i++) {
+    FallbackTableEntry* entry = _table[i]._next;
+    while (entry != nullptr) {
+      FallbackTableEntry* next = entry->_next;
+      FREE_C_HEAP_OBJ(entry);
+      entry = next;
+    }
+  }
+}
+
+uint FallbackTable::home_index(HeapWord* from) {
+  uint64_t val = reinterpret_cast<uint64_t>(from);
+  // This is the mixer stage of the murmur3 hashing:
+  // https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp
+  val ^= val >> 33;
+  val *= 0xff51afd7ed558ccdULL;
+  val ^= val >> 33;
+  val *= 0xc4ceb9fe1a85ec53ULL;
+  val ^= val >> 33;
+  // Shift to table-size.
+  val = val >> (64 - log2i_exact(TABLE_SIZE));
+  uint idx = static_cast<uint>(val);
+  assert(idx < TABLE_SIZE, "must fit in table: idx: %u, table-size: %u, table-size-bits: %d",
+         idx, TABLE_SIZE, log2i_exact(TABLE_SIZE));
+  return idx;
+}
+
+void FallbackTable::forward_to(HeapWord* from, HeapWord* to) {
+  uint idx = home_index(from);
+  FallbackTableEntry* head = &_table[idx];
+  FallbackTableEntry* entry = head;
+  // Search existing entry in chain starting at idx.
+  while (entry != nullptr) {
+    if (entry->_from == from) {
+      break;
+    }
+    entry = entry->_next;
+  }
+  if (entry == nullptr) {
+    // No entry found, create new one and insert after head.
+    FallbackTableEntry* new_entry = NEW_C_HEAP_OBJ(FallbackTableEntry, mtGC);
+    new_entry->_next = head->_next;
+    new_entry->_from = head->_from;
+    new_entry->_to   = head->_to;
```

Comment:
You could probably just use assignment here, which does memberwise copy. `*new_entry = *head;`